pkg/services/baremetal/host/host.go

```diff
@@ -1202,6 +1205,142 @@ func (s *Service) actionImageInstalling(ctx context.Context) actionResult {
+	}
+
+	switch state {
+	case sshclient.ImageURLCommandStateRunning:
+		conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+			string(host.Spec.Status.ProvisioningState), clusterv1.ConditionSeverityInfo,
+			"image-url-command still running")
+		return actionContinue{delay: 10 * time.Second}
+
+	case sshclient.ImageURLCommandStateFinishedSuccessfully:
+		conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+			string(host.Spec.Status.ProvisioningState), clusterv1.ConditionSeverityInfo,
+			"baremetal-image-url-command finished successfully. Rebooting now")
+
+		record.Event(s.scope.HetznerBareMetalHost, "ImageURLCommandOutput", logFile)
+		s.scope.Logger.Info("ImageURLCommandOutput", "logFile", logFile)
+
+		// Update name in robot API
+		if _, err := s.scope.RobotClient.SetBMServerName(s.scope.HetznerBareMetalHost.Spec.ServerID, s.scope.Hostname()); err != nil {
+			record.Warn(s.scope.HetznerBareMetalHost, "SetBMServerNameFailed", err.Error())
+			s.handleRobotRateLimitExceeded(err, "SetBMServerName")
+			return actionError{err: fmt.Errorf("failed to update name of host in robot API: %w", err)}
+		}
+
+		if err := sshClient.Reboot().Err; err != nil {
+			err = fmt.Errorf("failed to reboot server (after install-image): %w", err)
+			record.Warn(s.scope.HetznerBareMetalHost, "RebootFailed", err.Error())
+			return actionError{err: err}
+		}
+
+		msg := "machine image and cloud-init data got installed (via image-url-command)"
+		createSSHRebootEvent(ctx, s.scope.HetznerBareMetalHost, msg)
+		conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+			string(host.Spec.Status.ProvisioningState),
+			clusterv1.ConditionSeverityInfo,
+			"%s", msg)
+
+		// clear potential errors - all done
+		s.scope.HetznerBareMetalHost.ClearError()
+		return actionComplete{}
+
+	case sshclient.ImageURLCommandStateFailed:
+		record.Warn(s.scope.HetznerBareMetalHost, "InstallImageNotSuccessful", logFile)
+		msg := "image-url-command failed"
+		s.scope.Logger.Error(nil, msg, "logFile", logFile)
+		conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+			string(host.Spec.Status.ProvisioningState), clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return s.recordActionFailure(infrav1.FatalError, msg)
+
+	case sshclient.ImageURLCommandStateNotStarted:
+		data, err := s.scope.GetRawBootstrapData(ctx)
+		if err != nil {
+			return actionError{err: fmt.Errorf("baremetal GetRawBootstrapData failed: %w", err)}
+		}
+		if s.scope.ImageURLCommand == "" {
+			err = errors.New("internal error: --baremetal-image-url-command is not set?")
+			s.scope.Logger.Error(err, "")
+			conditions.MarkFalse(s.scope.HetznerBareMetalHost, infrav1.ProvisionSucceededCondition,
+				string(s.scope.HetznerBareMetalHost.Spec.Status.ProvisioningState),
+				clusterv1.ConditionSeverityWarning,
+				"%s", err.Error())
+			return actionError{err: err}
+		}
+		// get the information about storage devices again to have the latest names.
```

Comment:
this seems to be a new feature. Is that correct? Can you ideally make a separate PR out of this, to make it visible to someone else?

Reply:
@janiskemper no, this is not a new feature. I just added a comment, so that the reader understands why we do it.

Reply:
From what I understand, you add the function obtainHardwareDetailsStorage again here, even though we didn't do that before, because something could have changed in the meantime. This is independent of the topic of this PR from my understanding, because this would be something we would profit from in general. 

If I misunderstood this code, please explain me.

Reply:
@janiskemper obtainHardwareDetailsStorage gets called twice. It depends on the way the machine gets provisioned: with image-url-command or without it. This is all part of the topic of this PR.


first (existing) usage of obtainHardwareDetailsStorage (in createAutoSetupInput (without image-url-command)): https://github.com/syself/cluster-api-provider-hetzner/blob/4cdbf1ebdf3afe9b95ab7ca1aa11d4cf8b51dc5d/pkg/services/baremetal/host/host.go#L1600-L1610

second (new) usage of obtainHardwareDetailsStorage (in actionImageInstallingCustomImageURLCommand) : https://github.com/syself/cluster-api-provider-hetzner/blob/4cdbf1ebdf3afe9b95ab7ca1aa11d4cf8b51dc5d/pkg/services/baremetal/host/host.go#L1287-L1297

Reply:
I understood it now. I forgot about this feature that we added some time ago. Thanks!